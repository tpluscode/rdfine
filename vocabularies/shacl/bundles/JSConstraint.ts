import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import { JSExecutableMixin } from '../JSExecutable';
import { JSConstraintMixin } from '../JSConstraint';

export const JSConstraintBundle = [
  JSExecutableMixin as Mixin,
  JSConstraintMixin as Mixin];
