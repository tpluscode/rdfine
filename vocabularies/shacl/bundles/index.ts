export { AbstractResultBundle } from './AbstractResult';
export { ConstraintComponentBundle } from './ConstraintComponent';
export { FunctionBundle } from './Function';
export { JSConstraintBundle } from './JSConstraint';
export { JSExecutableBundle } from './JSExecutable';
export { JSFunctionBundle } from './JSFunction';
export { JSLibraryBundle } from './JSLibrary';
export { JSRuleBundle } from './JSRule';
export { JSTargetBundle } from './JSTarget';
export { JSTargetTypeBundle } from './JSTargetType';
export { JSValidatorBundle } from './JSValidator';
export { NodeShapeBundle } from './NodeShape';
export { ParameterBundle } from './Parameter';
export { ParameterizableBundle } from './Parameterizable';
export { PrefixDeclarationBundle } from './PrefixDeclaration';
export { PropertyGroupBundle } from './PropertyGroup';
export { PropertyShapeBundle } from './PropertyShape';
export { ResultAnnotationBundle } from './ResultAnnotation';
export { RuleBundle } from './Rule';
export { SeverityBundle } from './Severity';
export { ShapeBundle } from './Shape';
export { SPARQLAskExecutableBundle } from './SPARQLAskExecutable';
export { SPARQLAskValidatorBundle } from './SPARQLAskValidator';
export { SPARQLConstraintBundle } from './SPARQLConstraint';
export { SPARQLConstructExecutableBundle } from './SPARQLConstructExecutable';
export { SPARQLExecutableBundle } from './SPARQLExecutable';
export { SPARQLFunctionBundle } from './SPARQLFunction';
export { SPARQLRuleBundle } from './SPARQLRule';
export { SPARQLSelectExecutableBundle } from './SPARQLSelectExecutable';
export { SPARQLSelectValidatorBundle } from './SPARQLSelectValidator';
export { SPARQLTargetBundle } from './SPARQLTarget';
export { SPARQLTargetTypeBundle } from './SPARQLTargetType';
export { SPARQLUpdateExecutableBundle } from './SPARQLUpdateExecutable';
export { TargetBundle } from './Target';
export { TargetTypeBundle } from './TargetType';
export { TripleRuleBundle } from './TripleRule';
export { ValidationReportBundle } from './ValidationReport';
export { ValidationResultBundle } from './ValidationResult';
export { ValidatorBundle } from './Validator';
