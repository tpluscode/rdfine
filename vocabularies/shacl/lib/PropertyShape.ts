import RdfResourceImpl, { Constructor, namespace, RdfResource, property } from '@tpluscode/rdfine';
import * as $rdf from '@rdf-esm/data-model';
import type * as RDF from 'rdf-js';
import { sh } from './namespace';
import type { Initializer, ResourceNode } from '@tpluscode/rdfine/RdfResource';
import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import type * as Sh from '..';
import type * as Rdfs from '@rdfine/rdfs';
import type * as Rdf from '@rdfine/rdf';
import { ShapeMixin } from './Shape';
import { DatatypeMixin as RdfsDatatypeMixin } from '@rdfine/rdfs/lib/Datatype';
import { PropertyMixin as RdfPropertyMixin } from '@rdfine/rdf/lib/Property';
import { ResourceMixin as RdfsResourceMixin } from '@rdfine/rdfs/lib/Resource';

export interface PropertyShape<D extends RDF.DatasetCore = RDF.DatasetCore> extends Sh.Shape<D>, RdfResource<D> {
  datatype: Rdfs.Datatype<D> | undefined;
  defaultValue: RDF.Term | undefined;
  description: string | undefined;
  disjoint: Rdf.Property<D> | undefined;
  '_equals': Rdf.Property<D> | undefined;
  flags: string | undefined;
  group: Sh.PropertyGroup<D> | undefined;
  hasValue: Array<RDF.Term>;
  languageIn: Array<string>;
  lessThan: Rdf.Property<D> | undefined;
  lessThanOrEquals: Rdf.Property<D> | undefined;
  maxCount: number | undefined;
  maxExclusive: number | undefined;
  maxInclusive: number | undefined;
  maxLength: number | undefined;
  minCount: number | undefined;
  minExclusive: number | undefined;
  minInclusive: number | undefined;
  minLength: number | undefined;
  name: string | undefined;
  order: number | undefined;
  path: Rdfs.Resource<D> | Array<Rdfs.Resource<D>> | undefined;
  pattern: string | undefined;
  uniqueLang: boolean | undefined;
}

export function PropertyShapeMixin<Base extends Constructor>(Resource: Base): Constructor<PropertyShape> & Base {
  @namespace(sh)
  class PropertyShapeClass extends ShapeMixin(Resource) implements PropertyShape {
    @property.resource({ as: [RdfsDatatypeMixin] })
    datatype: Rdfs.Datatype | undefined;
    @property()
    defaultValue: RDF.Term | undefined;
    @property.literal()
    description: string | undefined;
    @property.resource({ as: [RdfPropertyMixin] })
    disjoint: Rdf.Property | undefined;
    @property.resource({ as: [RdfPropertyMixin] })
    '_equals': Rdf.Property | undefined;
    @property.literal()
    flags: string | undefined;
    @property.resource({ implicitTypes: [sh.PropertyGroup] })
    group: Sh.PropertyGroup | undefined;
    @property({ values: 'array' })
    hasValue!: Array<RDF.Term>;
    @property.literal({ values: 'list' })
    languageIn!: Array<string>;
    @property.resource({ as: [RdfPropertyMixin] })
    lessThan: Rdf.Property | undefined;
    @property.resource({ as: [RdfPropertyMixin] })
    lessThanOrEquals: Rdf.Property | undefined;
    @property.literal({ type: Number })
    maxCount: number | undefined;
    @property.literal({ type: Number })
    maxExclusive: number | undefined;
    @property.literal({ type: Number })
    maxInclusive: number | undefined;
    @property.literal({ type: Number })
    maxLength: number | undefined;
    @property.literal({ type: Number })
    minCount: number | undefined;
    @property.literal({ type: Number })
    minExclusive: number | undefined;
    @property.literal({ type: Number })
    minInclusive: number | undefined;
    @property.literal({ type: Number })
    minLength: number | undefined;
    @property.literal()
    name: string | undefined;
    @property.literal({ type: Number })
    order: number | undefined;
    @property.resource({ values: ['list', 'single'], as: [RdfsResourceMixin] })
    path: Rdfs.Resource | Array<Rdfs.Resource> | undefined;
    @property.literal()
    pattern: string | undefined;
    @property.literal({ type: Boolean })
    uniqueLang: boolean | undefined;
  }
  return PropertyShapeClass
}

class PropertyShapeImpl extends PropertyShapeMixin(RdfResourceImpl) {
  constructor(arg: ResourceNode, init?: Initializer<PropertyShape>) {
    super(arg, init)
    this.types.add(sh.PropertyShape)
  }

  static readonly __mixins: Mixin[] = [PropertyShapeMixin, ShapeMixin];
}
PropertyShapeMixin.appliesTo = sh.PropertyShape
PropertyShapeMixin.Class = PropertyShapeImpl
