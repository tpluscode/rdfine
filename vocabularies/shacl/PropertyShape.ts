import RdfResourceImpl, { Constructor, namespace, RdfResource, property } from '@tpluscode/rdfine';
import type * as RDF from 'rdf-js';
import { sh } from './lib/namespace';
import type { Initializer, ResourceNode } from '@tpluscode/rdfine/RdfResource';
import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import type * as Sh from '.';
import type * as Rdfs from '@rdfine/rdfs';
import { ResourceMixin as RdfsResourceMixin } from '@rdfine/rdfs/Resource';
import { ShapeMixin } from './Shape';

export interface PropertyShape extends Sh.Shape, RdfResource {
  defaultValue: RDF.Term;
  description: RDF.Term;
  group: Sh.PropertyGroup;
  name: RDF.Term;
  path: Rdfs.Resource;
}

export function PropertyShapeMixin<Base extends Constructor>(Resource: Base) {
  @namespace(sh)
  class PropertyShapeClass extends ShapeMixin(Resource) implements PropertyShape {
    @property()
    defaultValue!: RDF.Term;
    @property()
    description!: RDF.Term;
    @property.resource({ implicitTypes: [sh.PropertyGroup] })
    group!: Sh.PropertyGroup;
    @property()
    name!: RDF.Term;
    @property.resource({ as: [RdfsResourceMixin] })
    path!: Rdfs.Resource;
  }
  return PropertyShapeClass
}

class PropertyShapeImpl extends PropertyShapeMixin(RdfResourceImpl) {
  constructor(arg: ResourceNode, init?: Initializer<PropertyShape>) {
    super(arg, init)
    this.types.add(sh.PropertyShape)
  }

  static readonly __mixins: Mixin[] = [PropertyShapeMixin, ShapeMixin];
}
PropertyShapeMixin.appliesTo = sh.PropertyShape
PropertyShapeMixin.Class = PropertyShapeImpl
