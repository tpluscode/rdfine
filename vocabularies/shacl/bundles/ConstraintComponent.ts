import { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import { ParameterizableMixin } from '../Parameterizable';
import { ValidatorMixin } from '../Validator';
import { ConstraintComponentMixin } from '../ConstraintComponent';

export const ConstraintComponentBundle = [
  ParameterizableMixin as Mixin,
  ValidatorMixin as Mixin,
  ConstraintComponentMixin as Mixin];
