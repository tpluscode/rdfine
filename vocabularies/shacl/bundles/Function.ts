import { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import { ParameterizableMixin } from '../Parameterizable';
import { FunctionMixin } from '../Function';

export const FunctionBundle = [
  ParameterizableMixin as Mixin,
  FunctionMixin as Mixin];
