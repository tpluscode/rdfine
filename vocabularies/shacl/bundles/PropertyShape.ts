import type { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import { ShapeMixin } from '../Shape';
import { PropertyGroupMixin } from '../PropertyGroup';
import { PropertyShapeMixin } from '../PropertyShape';

export const PropertyShapeBundle = [
  ShapeMixin as Mixin,
  PropertyGroupMixin as Mixin,
  PropertyShapeMixin as Mixin];
