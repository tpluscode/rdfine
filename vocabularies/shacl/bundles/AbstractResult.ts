import { Mixin } from '@tpluscode/rdfine/lib/ResourceFactory';
import { AbstractResultMixin } from '../AbstractResult';
import { SeverityMixin } from '../Severity';
import { ConstraintComponentMixin } from '../ConstraintComponent';
import { ShapeMixin } from '../Shape';

export const AbstractResultBundle = [
  AbstractResultMixin as Mixin,
  SeverityMixin as Mixin,
  ConstraintComponentMixin as Mixin,
  ShapeMixin as Mixin];
