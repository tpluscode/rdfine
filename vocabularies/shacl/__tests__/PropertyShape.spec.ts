import clownface from 'clownface'
import $rdf from 'rdf-ext'
import RDF from '@rdfjs/data-model'
import RdfResource from '@tpluscode/rdfine'
import { rdfs, skos } from '@tpluscode/rdf-ns-builders'
import { ResourceMixin } from '@rdfine/rdfs/Resource'
import { PropertyShape, PropertyShapeMixin } from '../PropertyShape';
import { PropertyShapeBundle } from '../bundles'

RdfResource.factory.addMixin(...PropertyShapeBundle)
RdfResource.factory.addMixin(ResourceMixin)

describe('PropertyShape', () => {
  describe('initializer', () => {
    it('lets initializing path with named node', () => {
      // given
      const dataset = $rdf.dataset()
      const graph = clownface({ dataset, term: RDF.namedNode('http://example.com/shape') })

      // when
      // eslint-disable-next-line no-new
      new PropertyShapeMixin.Class(graph, {
        path: rdfs.label,
      })

      // then
      expect(dataset.toCanonical()).toMatchSnapshot()
    })

    it('lets initializing path with named node array', () => {
      // given
      const dataset = $rdf.dataset()
      const graph = clownface({ dataset, term: RDF.namedNode('http://example.com/shape') })

      // when
      // eslint-disable-next-line no-new
      new PropertyShapeMixin.Class(graph, {
        path: [skos.broader, rdfs.label],
      })

      // then
      expect(dataset.toCanonical()).toMatchSnapshot()
    })
  })
})
