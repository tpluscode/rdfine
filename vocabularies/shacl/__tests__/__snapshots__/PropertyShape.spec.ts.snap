// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyShape initializer lets initializing path with named node array: toMatchSnapshot 1`] = `
"<http://example.com/shape> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/shacl#PropertyShape> .
<http://example.com/shape> <http://www.w3.org/ns/shacl#path> _:c14n1 .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> <http://www.w3.org/2000/01/rdf-schema#label> .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> <http://www.w3.org/2004/02/skos/core#broader> .
_:c14n1 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> _:c14n0 .
"
`;

exports[`PropertyShape initializer lets initializing path with named node: toMatchSnapshot 1`] = `
"<http://example.com/shape> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/shacl#PropertyShape> .
<http://example.com/shape> <http://www.w3.org/ns/shacl#path> _:c14n0 .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#first> <http://www.w3.org/2000/01/rdf-schema#label> .
_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#rest> <http://www.w3.org/1999/02/22-rdf-syntax-ns#nil> .
"
`;
